src/lib/disk_cache/test/test_cache_deadlock.ml

```diff
@@ -0,0 +1,141 @@
+
+  (* We convert to/from string for bin_io, injecting our GC trigger during conversion *)
+  include
+    Binable.Of_binable_without_uuid
+      (String)
+      (struct
+        type nonrec t = t
+
+        let to_binable t =
+          (* Trigger GC during serialization if requested *)
+          Option.iter t.trigger_gc ~f:(fun f -> f ()) ;
+          t.value
+
+        let of_binable value = { value; trigger_gc = None }
+      end)
+end
+
+module type Cache_intf = sig
+  module Make : Disk_cache_intf.F
+end
+
+let run_test_with_cache (module Cache_impl : Cache_intf) ~timeout_seconds ~tmpdir =
+  let module Cache = Cache_impl.Make (Evil_data) in
+  let open Deferred.Let_syntax in
+  let%bind cache_result = Cache.initialize tmpdir ~logger:(Logger.null ()) in
+  let cache =
+    match cache_result with
+    | Ok cache ->
+        cache
+    | Error (`Initialization_error err) ->
+        failwithf "Failed to initialize cache: %s" (Error.to_string_hum err) ()
+  in
+
+  (* Create a cache entry that will be garbage collected *)
+  let entry_ref = ref None in
+  let entry_data = { Evil_data.value = "entry"; trigger_gc = None } in
+  entry_ref := Some (Cache.put cache entry_data) ;
+
+  (* Create evil data that triggers GC during serialization *)
+  let evil_data =
+    { Evil_data.value = "evil"
+    ; trigger_gc =
+        Some
+          (fun () ->
+            (* Clear reference and trigger GC to run finalizers *)
+            entry_ref := None ;
+            Core.printf "References cleared, triggering GC...\n%!" ;
+            Gc.compact () ;
+            Core.printf
+              "GC triggered during serialization - finalizers should run now\n%!" )
+    }
+  in
+
+  (* This put may deadlock if a finalizer runs during serialization *)
+  Core.printf "Attempting evil put...\n%!" ;
+
+  (* Run the potentially deadlocking operation with a timeout *)
+  let put_with_timeout () =
+    let put_deferred =
+      In_thread.run (fun () ->
+          ignore (Cache.put cache evil_data : Cache.id) ;
+          `Success )
+    in
+    match timeout_seconds with
+    | Some timeout ->
+        let timeout_span = Core.Time.Span.of_sec timeout in
+        Clock.with_timeout timeout_span put_deferred
+    | None ->
+        let%map result = put_deferred in
+        `Result result
+  in
+
+  let%bind result = put_with_timeout () in
+
+  match result with
+  | `Timeout ->
+      Core.printf
+        "\nDEADLOCK DETECTED: Cache.put timed out after %.1f seconds!\n%!"
+        (Option.value_exn timeout_seconds) ;
+      Core.printf "This indicates a deadlock in the cache implementation.\n%!" ;
+      failwith
+        "Cache deadlock detected - finalizer likely tried to acquire a lock during GC"
+  | `Result `Success ->
+      Core.printf "Evil put completed successfully (no deadlock)\n%!" ;
+      Core.printf "\nSUCCESS: Cache does NOT deadlock with finalizers.\n%!" ;
+      return ()
+
+let make_command (module Cache_impl : Cache_intf) =
+  Command.async
```

Comment:
Let's drop the command boilerplate.

Let's make it a standalone dune test. If it is useful to sometimes specify or drop timeout and directory parameters, let's just resolve to environment variables maybe?

By default, this should be run as tests, so I don't think that making them executables is really helpful.